src/Microsoft.OpenApi.Hidi/OpenApiService.cs

```diff
@@ -115,6 +131,54 @@ public static class OpenApiService
             textWriter.Flush();
         }
 
+        /// <summary>
+        /// Converts CSDL to OpenAPI
+        /// </summary>
+        /// <param name="csdl">The CSDL stream.</param>
+        /// <returns>An OpenAPI document.</returns>
+        public static OpenApiDocument ConvertCsdlToOpenApi(Stream csdl)
+        {
+            using var reader = new StreamReader(csdl);
+            var csdlText = reader.ReadToEndAsync().GetAwaiter().GetResult();            
+            var edmModel = CsdlReader.Parse(XElement.Parse(csdlText).CreateReader());
+
+            var settings = new OpenApiConvertSettings()
+            {
+                EnableKeyAsSegment = true,
+                EnableOperationId = true,
+                PrefixEntityTypeNameBeforeKey = true,
+                TagDepth = 2,
+                EnablePagination = true,
+                EnableDiscriminatorValue = false,
+                EnableDerivedTypesReferencesForRequestBody = false,
+                EnableDerivedTypesReferencesForResponses = false,
+                ShowRootPath = true,
+                ShowLinks = true
+            };
+            OpenApiDocument document = edmModel.ConvertToOpenApi(settings);
+
+            document = FixReferences(document);
+
+            return document;
+        }
+
+        /// <summary>
+        /// Fixes the references in the resulting OpenApiDocument.
+        /// </summary>
+        /// <param name="document"> The converted OpenApiDocument.</param>
+        /// <returns> A valid OpenApiDocument instance.</returns>
+        public static OpenApiDocument FixReferences(OpenApiDocument document)
+        {
+            // This method is only needed because the output of ConvertToOpenApi isn't quite a valid OpenApiDocument instance.
```

Comment:
is there a plan to fix the underlying issue in the conversion?

Reply:
There is a ticket that has been created in the OpenAPI.NET.OData conversion lib to address this issue